src/uct/ib/base/ib_md.c

```diff
@@ -1694,33 +1705,79 @@ ucs_status_t uct_ib_md_open_common(uct_ib_md_t *md,
 
     return UCS_OK;
 
-err_dealloc_pd:
-    ibv_dealloc_pd(md->pd);
 err_cleanup_device:
     uct_ib_device_cleanup(&md->dev);
 err_release_stats:
     UCS_STATS_NODE_FREE(md->stats);
+err_device_config_release:
+    uct_ib_md_release_device_config(md);
 err:
     return status;
 }
 
-void uct_ib_md_close(uct_md_h uct_md)
+void uct_ib_md_close_common(uct_ib_md_t *md)
 {
-    uct_ib_md_t *md = ucs_derived_of(uct_md, uct_ib_md_t);
-
-    /* Must be done before md->ops->cleanup, since it can call functions from
-     * md->ops */
     uct_ib_md_release_reg_method(md);
-    md->ops->cleanup(md);
-    uct_ib_md_release_device_config(md);
-    uct_ib_device_cleanup_ah_cached(&md->dev);
-    ibv_dealloc_pd(md->pd);
     uct_ib_device_cleanup(&md->dev);
-    ibv_close_device(md->dev.ibv_context);
     UCS_STATS_NODE_FREE(md->stats);
+    uct_ib_md_release_device_config(md);
+}
+
+void uct_ib_md_device_context_close(struct ibv_context *ctx)
+{
+    int ret = ibv_close_device(ctx);
+    if (ret != 0) {
+        ucs_warn("ibv_close_device() failed: %m");
+    }
+}
+
+uct_ib_md_t* uct_ib_md_alloc(size_t size, const char *name,
+                             struct ibv_context *ctx)
+{
+    uct_ib_md_t *md;
+
+    md = ucs_calloc(1, size, name);
+    if (md == NULL) {
+        ucs_error("failed to allocate memory for md");
+        goto err;
+    }
+
+    md->dev.ibv_context = ctx;
+    md->pd              = ibv_alloc_pd(md->dev.ibv_context);
+    if (md->pd == NULL) {
+        ucs_error("ibv_alloc_pd() failed: %m");
+        goto err_md_free;
+    }
+
+    return md;
+
+err_md_free:
+    ucs_free(md);
+err:
+    return NULL;
+}
+
+void uct_ib_md_free(uct_ib_md_t *md)
```

Comment:
maybe just remove this wrapper? and not print a warning if fails

Reply:
better to print errors, it could help finding potential problems with reordering operations during cleanup